File: src/main/java/com/medisphere/service/HealthTwinService.java
package com.medisphere.service;

import com.medisphere.domain.HealthTwin;
import com.medisphere.domain.Patient;
import com.medisphere.repository.HealthTwinRepository;
import com.medisphere.repository.PatientRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@Slf4j
public class HealthTwinService {

    @Autowired
    private HealthTwinRepository healthTwinRepository;

    @Autowired
    private PatientRepository patientRepository;

    @Autowired
    private AuditService auditService;

    @Transactional
    public HealthTwin createHealthTwin(String patientId) {
        log.info("Creating health twin for patient: {}", patientId);

        Optional<Patient> optionalPatient = patientRepository.findById(patientId);
        if (optionalPatient.isEmpty()) {
            throw new IllegalArgumentException("Patient not found: " + patientId);
        }

        Patient patient = optionalPatient.get();

        HealthTwin twin = HealthTwin.builder()
                .twinId(UUID.randomUUID().toString())
                .patientId(patientId)
                .firstName(patient.getFirstName())
                .lastName(patient.getLastName())
                .dateOfBirth(patient.getDateOfBirth())
                .gender(patient.getGender())
                .modelVersion("1.0.0")
                .createdAt(LocalDateTime.now())
                .lastUpdated(LocalDateTime.now())
                .twinDataCompleteness(0)
                .overallRiskLevel("UNKNOWN")
                .hipaaCompliant(true)
                .build();

        HealthTwin savedTwin = healthTwinRepository.save(twin);
        auditService.logAction("HEALTH_TWIN_CREATED", savedTwin.getId(), patient.getFirstName());
        log.info("Health twin created: {}", savedTwin.getId());

        return savedTwin;
    }

    public Optional<HealthTwin> getHealthTwinByPatientId(String patientId) {
        log.debug("Fetching health twin for patient: {}", patientId);
        return healthTwinRepository.findByPatientId(patientId);
    }

    @Transactional
    public HealthTwin updateHealthTwin(HealthTwin twin) {
        log.info("Updating health twin: {}", twin.getId());
        twin.setLastUpdated(LocalDateTime.now());
        twin.setTwinDataCompleteness(twin.calculateDataCompleteness());

        HealthTwin updatedTwin = healthTwinRepository.save(twin);
        auditService.logAction("HEALTH_TWIN_UPDATED", twin.getId(), twin.getFirstName());

        return updatedTwin;
    }

    public List<HealthTwin> getHighRiskPatients() {
        log.debug("Fetching high-risk patients");
        return healthTwinRepository.findHighRiskPatients();
    }

    public List<HealthTwin> getTwinsByCompleteness(int minimumCompleteness) {
        log.debug("Fetching twins with minimum completeness: {}%", minimumCompleteness);
        return healthTwinRepository.findTwinsByMinimumCompleteness(minimumCompleteness);
    }

    public List<HealthTwin> getTwinsWithActiveAlerts() {
        log.debug("Fetching twins with active alerts");
        return healthTwinRepository.findTwinsWithActiveAlerts();
    }

    @Transactional
    public void updateRiskScore(String twinId, String riskType, float score) {
        Optional<HealthTwin> optionalTwin = healthTwinRepository.findById(twinId);
        if (optionalTwin.isPresent()) {
            HealthTwin twin = optionalTwin.get();
            twin.getRiskScores().put(riskType, score);
            updateOverallRiskLevel(twin);
            healthTwinRepository.save(twin);
            log.info("Risk score updated for twin {}: {} = {}", twinId, riskType, score);
        }
    }

    private void updateOverallRiskLevel(HealthTwin twin) {
        if (twin.getRiskScores() == null || twin.getRiskScores().isEmpty()) {
            twin.setOverallRiskLevel("UNKNOWN");
            return;
        }

        float maxRisk = twin.getRiskScores().values().stream()
                .max(Float::compare)
                .orElse(0.0f);

        if (maxRisk >= 0.7f) {
            twin.setOverallRiskLevel("CRITICAL");
        } else if (maxRisk >= 0.5f) {
            twin.setOverallRiskLevel("HIGH");
        } else if (maxRisk >= 0.3f) {
            twin.setOverallRiskLevel("MODERATE");
        } else {
            twin.setOverallRiskLevel("LOW");
        }
    }

    @Transactional
    public void addAlert(String twinId, HealthTwin.Alert alert) {
        Optional<HealthTwin> optionalTwin = healthTwinRepository.findById(twinId);
        if (optionalTwin.isPresent()) {
            HealthTwin twin = optionalTwin.get();
            if (twin.getActiveAlerts() == null) {
                twin.setActiveAlerts(new java.util.ArrayList<>());
            }
            twin.getActiveAlerts().add(alert);
            healthTwinRepository.save(twin);
            log.info("Alert added to twin {}: {}", twinId, alert.getMessage());
        }
    }

    public long countHighRiskPatients() {
        return healthTwinRepository.countByOverallRiskLevel("HIGH") +
               healthTwinRepository.countByOverallRiskLevel("CRITICAL");
    }

    public List<HealthTwin> getAllHealthTwins() {
        log.debug("Fetching all health twins");
        return healthTwinRepository.findAll();
    }

    public long countTotalTwins() {
        return healthTwinRepository.count();
    }
}

File: frontend/src/app/services/health-twin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface HealthTwin {
  id?: string;
  twinId?: string;
  patientId: string;
  firstName: string;
  lastName: string;
  modelVersion?: string;
  currentVitals?: any;
  riskScores?: { [key: string]: number };
  overallRiskLevel?: string;
  twinDataCompleteness?: number;
  activeAlerts?: any[];
  hipaaCompliant?: boolean;
  lastUpdated?: string;
}

@Injectable({
  providedIn: 'root'
})
export class HealthTwinService {
  private apiUrl = 'http://localhost:8080/api/v1/health-twins';

  constructor(private http: HttpClient) { }

  createHealthTwin(patientId: string): Observable<HealthTwin> {
    return this.http.post<HealthTwin>(`${this.apiUrl}/patient/${patientId}`, {});
  }

  getHealthTwin(patientId: string): Observable<HealthTwin> {
    return this.http.get<HealthTwin>(`${this.apiUrl}/patient/${patientId}`);
  }

  updateHealthTwin(twinId: string, twin: HealthTwin): Observable<HealthTwin> {
    return this.http.put<HealthTwin>(`${this.apiUrl}/${twinId}`, twin);
  }

  getHighRiskPatients(): Observable<HealthTwin[]> {
    return this.http.get<HealthTwin[]>(`${this.apiUrl}/list/high-risk`);
  }

  getTwinsByCompleteness(minCompleteness: number): Observable<HealthTwin[]> {
    return this.http.get<HealthTwin[]>(`${this.apiUrl}/list/completeness/${minCompleteness}`);
  }

  countHighRiskPatients(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/stats/high-risk-count`);
  }

  countTotalTwins(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/stats/total-count`);
  }

  updateRiskScore(twinId: string, riskType: string, score: number): Observable<void> {
    return this.http.post<void>(
      `${this.apiUrl}/${twinId}/risk-score?riskType=${riskType}&score=${score}`,
      {}
    );
  }
}
